Prepare to 0.2.0 release

AnnaShaleva opened the pull request:
Depends on #230, #166.

Close #169, close #226, close #224, close #231, close #175.

This PR needs to be merged before T4-compatible release.

AnnaShaleva committed:
contracts: add Treasury systemcontract documentation

Close #226.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
config: add native Treasury to genesis allocations

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
Makefile: remove privnet generation code

Our privnet settings are static. we'll never change them. Also, genesis
templates are removed so that privnet configuration update should be
done only for the genesis files themselves.

Close #169.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
config: adjust testnet configuration for T4

Change the magic to 0xBA9304, add NeoXBurn schedule.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
systemcontracts: remove PolicyABI global variable

It's unused and out-of-date, thus I propose to remove it and introduce
it back only if we really need it in the code.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
systemcontracts: add missing native proxy hashes

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
contracts: upgrade solidity compiler target

Shanghai is the latest supported hardfork for Neo X. Further updates
should be bound to Dencun in #189.

Close #224.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
params: move to v0.2.0 release cycle

Skip v0.1.2, it won't be released. Bump the minor version since it
contains a lot of changes incompatible with v0.1.1 version.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
README: add basic docs about starting Neo X node

Close #175.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
systemcontracts, contracts: reserve 11 system contracts

One is for DKG contract, the rest 10 are for possible further usage.
Reserve addresses in genesis allocations and create a stub
implementation in system contracts folder.

Close #231.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented on config/genesis_testnet.json:
```diff
@@ -85,7 +85,7 @@
     },
     "0x1212000000000000000000000000000000000004": {
-      "balance": "60000000000000000000000000",
+      "balance": "0", // TODO: do we need at least *some* initial balance for native BridgeProxy?
```
@mialbu, @gsmachado, @steven1227 for T4 all native Bridge contract funds are moved to native Treasury contract. Do we need to supply Bridge contract with at least *some* initial amount of GAS in the genesis allocations?

mialbu replied:
Imo there should be some initial allocation to the bridge. Otherwise, we'd require a governance vote before the bridge can be used.

AnnaShaleva replied:
Yes, it's the same chicken-and-egg problem. So how much funds do you think we should give to the bridge?

gsmachado replied:
I'd say we need just a bit of GAS for this. Like, 50 GAS to be on the safe side? wdyt?

steven1227 replied:
Maybe a tiny amount is enough ?

gsmachado replied:
Let's set 5 GAS, then. Or, I'm also fine with 1 GAS, if it's enough.

@mialbu will this be more than enough for the initial steps, right?

mialbu replied:
Actually, if it's a tiny amount, we need to fund it anyway rather soon already. Then I'm not sure if there's any point in allocating any funds to the bridge in the genesis.

AnnaShaleva replied:
Initial transfer from Treasury (where all Bridge funds are stored) requires votes from CN accounts. CN accounts should be supplied with GAS before they can vote. We don't put genesis allocations for CN accounts, they are supposed to bridge initial amount of funds to themselves. So the question is: can we perform these initial bridging to CN accounts with no GAS on the Bridge contract balance? I suppose we can't, that's why we need at least *some* amount for this genesis allocation.

mialbu replied:
Oh okay, I wasn't aware of that. Yes, in that case the bridge needs some initial allocation, absolutely.

There is a minimum deposit amount on the N3 side. It's been 1 Gas so far, but it can be modified as well. So, the initial allocation should be at least that amount, and it also needs to be enough to cover all the fees for the necessary transactions to fund the bridge. If 1 Gas is sufficient for all required transactions, then 1 Gas is sufficient as initial allocation to the bridge.

AnnaShaleva replied:
1GAS for BridgeProxy contract is allocated.

AnnaShaleva commented on config/genesis_testnet.json:
```diff
@@ -121,6 +121,13 @@
       }
     },
+    "0x1212000000000000000000000000000000000006": {
+      "balance": "60000000000000000000000000", // TODO: need more?
```
@mialbu, @gsmachado, @steven1227 what is the desired balance of native Treasury contract for T4?

AnnaShaleva replied:
As discussed, changed to `100,000,000`.

AnnaShaleva commented on config/genesis_testnet.json:
```diff
@@ -121,6 +121,13 @@
+      "balance": "60000000000000000000000000", // TODO: need more?
+      "code": "0x6080604052600a600c565b005b60186014601a565b6051565b565b6000604c7f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc546001600160a01b031690565b905090565b3660008037600080366000845af43d6000803e808015606f573d6000f35b3d6000fdfea26469706673582212205473baeda08227b037acf1d10bff816015a11f298917e2465da958f40bcefac664736f6c63430008180033",
+      "storage": {
+        // TODO: fill the storage
```
@chenquanyu, @txhsl, @mialbu I'm not sure how to get the initial storage items for Treasury proxy contract. Could you please, explain or provide them?

txhsl replied:
Only a key `0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc` with its value `0x1212100000000000000000000000000000000006`.

AnnaShaleva replied:
Implementation storage item is added. @mialbu, could you please confirm that no additional storage items are needed?

AnnaShaleva replied:
UPD: Treasury is not upgradeable, thus we use a single genesis allocation for it with `0x1212000000000000000000000000000000000006` address and an empty storage items set.

mialbu replied:
I'm fixing some things from the audit still. It'd be great if we can use that new contract for the start of T4. The medium issues are already fixed (unmerged PRs), and there's one or two low severity issues that I'll implement tomorrow. After that I'll need to have another look at the storage entries anyway.

AnnaShaleva replied:
It's OK, just write back once the contracts are in their final state.

AnnaShaleva commented on config/genesis_testnet.json:
```diff
@@ -121,6 +121,13 @@
     },
+    "0x1212000000000000000000000000000000000006": {
+      "balance": "60000000000000000000000000", // TODO: need more?
+      "code": "0x6080604052600a600c565b005b60186014601a565b6051565b565b6000604c7f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc546001600160a01b031690565b905090565b3660008037600080366000845af43d6000803e808015606f573d6000f35b3d6000fdfea26469706673582212205473baeda08227b037acf1d10bff816015a11f298917e2465da958f40bcefac664736f6c63430008180033",
```
@chenquanyu, @txhsl all our native contracts are upgradable and supposed to have the same Proxy code, right? I noticed that some of our Proxy contracts have different code than the others. It's probably due to different compiler versions or something like that, but we need to unify this point. What is the desired code for the Proxy contracts?

txhsl replied:
Codes for addresses from `0x1212000000000000000000000000000000000001` to `0x1212000000000000000000000000000000000003` are provided by @chenquanyu and me, while from `0x1212000000000000000000000000000000000004` to `0x1212000000000000000000000000000000000005` are provided by AxLabs, so they could be different.

From my point of view, I will recommend the code of `0x1212000000000000000000000000000000000001`.

AnnaShaleva replied:
> I will recommend the code of 0x1212000000000000000000000000000000000001

Fixed, let's use it for new contracts.

AnnaShaleva commented on config/genesis_testnet.json:
```diff
@@ -140,6 +147,10 @@
+    },
+    "0x1212100000000000000000000000000000000006": {
+      "balance": "0",
+      "code": "0x608060405234801561000f575f80fd5b506004361061003f575f3560e01c80630558839d14610043578063a5257a1d14610058578063cea107b71461008f575b5f80fd5b61005661005136600461055e565b6100aa565b005b61007373121200000000000000000000000000000000000381565b6040516001600160a01b03909116815260200160405180910390f35b61007373121200000000000000000000000000000000000481565b60408051602081018390527fdd6d322687f552c3 ...
```
@chenquanyu, @txhsl, @mialbu could you please check that this Treasury implementation code matches the expectations? I compiled it using our hardhat config from `./contracts/hardhat.config.ts` with `evmVersion: "shanghai"`.

txhsl replied:
I'm not sure, but it is somehow different with my deploy code. @chenquanyu Can you check again as well?

AnnaShaleva replied:
I used the updated hardhat config from this PR to compile:
```
import { HardhatUserConfig } from "hardhat/config";
import "@nomicfoundation/hardhat-toolbox";

const config: HardhatUserConfig = {
  solidity: {
    version: "0.8.25",
    settings: {
      evmVersion: "shanghai",
      optimizer: {
        enabled: true,
        runs: 500,
      },
    },
  },
  paths: {
    sources: "./solidity",
    tests: "./test",
  }
};

export default config;
```

Then did `npx hardhat compile` and then took the code from `deployedBytecode` field of `go-ethereum/contracts/artifacts/solidity/Treasury.sol/Treasury.json`. Am I doing something wrong?

txhsl replied:
No, not. I reset the env and get the same result as this line.

AnnaShaleva commented on config/genesis_testnet.json:
```diff
@@ -128,6 +128,34 @@
+      "balance": "0",
+      "code": "0x6080604052600a600c565b005b60186014601a565b6051565b565b6000604c7f360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc546001600160a01b031690565b905090565b3660008037600080366000845af43d6000803e808015606f573d6000f35b3d6000fdfea26469706673582212205473baeda08227b037acf1d10bff816015a11f298917e2465da958f40bcefac664736f6c63430008180033",
+      "storage": {
+        // TODO
```
@chenquanyu, @txhsl I need your help in storage allocations for systemcontracts stubs. Also, please check that Proxy/implementation code matches the expectations.

txhsl replied:
The proxy code should be the same as other proxies', and the storage allocation needs only a key `0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc` with its corresponding implementation address, just like the Treasury allocates.

AnnaShaleva replied:
Fixed, could you check please?

txhsl replied:
LGTM.

AnnaShaleva commented on contracts/solidity/interfaces/IStub0.sol:
```diff
@@ -0,0 +1,8 @@
+interface IStub0 {
+
+    // a stub function
+    function stub() external;
```
@txhsl, @chenquanyu is this contract a valid Eth contract? We need to ensure that there will be no problem once we need to upgrade it.

txhsl replied:
It is, but to make the upgrade possible, we need it to inherit `GovProxyUpgradeable.sol`, and contain the following codes. :)

https://github.com/bane-labs/go-ethereum/blob/7017f7aa5fe678b90496564ebb0d155d51095976/contracts/solidity/Governance.sol#L15

And.

https://github.com/bane-labs/go-ethereum/blob/7017f7aa5fe678b90496564ebb0d155d51095976/contracts/solidity/Governance.sol#L69-L87

The `SELF` address is the corresponding implementation address, which is **different** in each implementation file.

chenquanyu replied:
Maybe we can write a new contract to store `SELF` in a specific storage key. Then we only need one contract for all the empty upgradable system contracts.

txhsl replied:
These codes and value is temporary usage, because UUPS proxy can't get correctly initialized if allocated in genesis file. After the first upgrade, new implementation contracts will get `natural` addresses and don't need these any more.

AnnaShaleva replied:
> to make the upgrade possible, we need it to inherit GovProxyUpgradeable.sol

Oh gosh, I was just copy-pasting the Treasury contract and completely forgot that Treasury is not upgradeable. Fixed, @txhsl, could you please check one more time?

txhsl replied:
Yeh, if Treasury is not upgradeable, then things are similar with GovProxyAdmin at `0x1212000000000000000000000000000000000000`. There will not be an implementation address, but only the implementation code in `0x1212000000000000000000000000000000000006`, and no need to allocate `0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc`.

txhsl replied:
BTW, we have updated nearly most of the contracts (including the bridge), due to the audit and some new features, so their implementation codes need to be compiled and updated again as well.

AnnaShaleva replied:
> Yeh, if Treasury is not upgradeable, then things are similar with GovProxyAdmin

Fixed, check it please.

> so their implementation codes need to be compiled and updated again as well.

Implementations are updated for GovernanceProxyAdmin, Governance, Policy and GovernanceReward contracts. But storage items are still not updated, I need your help to update storage items.

To update Bridge and BridgeManagement implementaitons I need the final version of these contracts for T4 from @mialbu.

mialbu replied:
> To update Bridge and BridgeManagement implementaitons I need the final version of these contracts for T4 from @mialbu.

https://github.com/bane-labs/go-ethereum/pull/236#discussion_r1662813824

txhsl replied:
>  But storage items are still not updated, I need your help to update storage items.

Since I made the contract storage compatible, we only need to add a new field for `candidateLimit`. Also, @steven1227 wants to increase both the `baseFee` and `minGasTipCap` to `20 Gwei`.
```json
"0x1212000000000000000000000000000000000002": {
      "storage": {
        "0x0000000000000000000000000000000000000000000000000000000000000002": "0x04a817c800",
        "0x0000000000000000000000000000000000000000000000000000000000000003": "0x04a817c800",
        "0x0000000000000000000000000000000000000000000000000000000000000004": "0x07d0",
        "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc": "0x1212100000000000000000000000000000000002"
      }
    },
```

AnnaShaleva replied:
Updated. And just to clarify: we don't need to change storage items for Governance or GovernanceVote contracts, right? There were changes in these contracts, but these changes did not affect the initializable variables.

txhsl replied:
> There were changes in these contracts, but these changes did not affect the initializable variables.

Yes.

AnnaShaleva reviewed (commented):
Please, consider answering these questions, it will help to finish this PR.

AnnaShaleva committed:
config: update native implementation in genesis config

Update GovernanceProxyAdmin, Governance, Policy and GovernanceReward
implementation bytecodes.

Update Policy genesis allocations:
  * increase baseFee and minGasTip to 20 Gwei
  * add candidateLimit storage item

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

chenquanyu commented on contracts/solidity/interfaces/IStub1.sol:
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.25;
+
+interface IStub1 {
```
Why do we need interface and method for stub contracts?

AnnaShaleva replied:
It's just a template for further extension, but if you'd like then I can remove interface and remove this method from this contract. Agree?

chenquanyu replied:
Maybe we don't need so many new empty implementation contracts, we can just point all these proxy contracts to the same implementation contract.

AnnaShaleva replied:
> I can remove interface and remove this method from this contract

Done.

> we can just point all these proxy contracts to the same implementation contract.

Can we? Implementation contract should be able to control the upgrade and all implementation contracts have different `SELF`:
```
address public constant SELF = 0x1212100000000000000000000000000000000009;
```

AnnaShaleva replied:
And I don't think that it's correct to point to the shared implementation from different proxies because we may have problems once we're going to update the implementation for one of the proxies. It's not quite clear to me what will happen with the rest of proxies in this case.

Could you please clarify? May be I'm wrong.

chenquanyu replied:
> And I don't think that it's correct to point to the shared implementation from different proxies because we may have problems once we're going to update the implementation for one of the proxies. It's not quite clear to me what will happen with the rest of proxies in this case.
> 
> Could you please clarify? May be I'm wrong.

Nothing will happen to other proxies  when you upgrade one proxy, you just change the implementation slot value when upgrading.

AnnaShaleva replied:
@txhsl, agree to use a single stub *implementation*?

AnnaShaleva replied:
@chenquanyu fixed, please check.

txhsl replied:
> @txhsl, agree to use a single stub _implementation_?

I'm OK with that if it works.

AnnaShaleva replied:
@chenquanyu could you please check that at least one stub can be correctly upgraded on privnet?

chenquanyu replied:
> @chenquanyu could you please check that at least one stub can be correctly upgraded on privnet?

I have checked that all stub contracts can be upgraded correctly on privnet. Maybe we should add more stub contracts to meet future needs.

AnnaShaleva replied:
What is the desired number? Do you have exact contract examples that should be added in future (except DKG)?

AnnaShaleva replied:
OK, how about 10 stub contracts? It's a lot, and I think we can fit our future needs into these contracts for sure.

txhsl replied:
Yes, it's a lot. Or just from `0x1212000000000000000000000000000000000000` to `0x121200000000000000000000000000000000000f`.

AnnaShaleva replied:
Added.

AnnaShaleva committed:
config: increase candidate registration fee

Increase candidate registration fee from 1000 GAS to 20000 GAS.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented on config/genesis_testnet.json:
```diff
@@ -47,7 +47,7 @@
         "0x0000000000000000000000000000000000000000000000000000000000000002": "0x0de0b6b3a7640000",
         "0x0000000000000000000000000000000000000000000000000000000000000003": "0x027b46536c66c8e3000000",
-        "0x0000000000000000000000000000000000000000000000000000000000000004": "0x3635c9adc5dea00000",
+        "0x0000000000000000000000000000000000000000000000000000000000000004": "0x43c33c1937564800000",
```
@txhsl, registration fee is set to 20K GAS, please check it's the right storage item.

txhsl replied:
It is the expected key and number value, but for a correct format without error, we may need `0x043c33c1937564800000` here.

txhsl replied:
If this value changes before launch, I will let @steven1227 inform here directly.

AnnaShaleva replied:
> but for a correct format without error, we may need 0x043c33c1937564800000 here.

Fixed.

mialbu committed:
config: add bridge data to genesis config for T4

mialbu committed:
config: update bridge data in privnet genesis files for T4

AnnaShaleva committed:
config: update GovProxyAdmin code

Replace this version with the one compiled from fresh sources.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@txhsl @chenquanyu waiting for you approval before the merge, Roman is on vacation.
